server/src/user/model.rs

```diff
@@ -20,34 +20,55 @@ pub struct UserData {
     pub password: String,
 }
 
-impl User {
-    fn generate_password(clear_password: &str) -> String {
-        bcrypt::hash(clear_password).unwrap()
-    }
+#[derive(Debug)]
+pub enum UserError {
+    UserNotFound,
+    UsernameTaken,
+    DatabaseError,
+    GenericError,
+}
 
-    pub fn find_all(conn: &PgConnection) -> Result<Vec<Self>, diesel::result::Error> {
+impl User {
+    pub fn find_all(conn: &PgConnection) -> Result<Vec<Self>, UserError> {
         use crate::schema::users::dsl::*;
 
         let items = users.load::<User>(conn)?;
         Ok(items)
     }
 
-    pub fn find(conn: &PgConnection, user_id: Uuid) -> Result<Option<Self>, diesel::result::Error> {
+    pub fn find(conn: &PgConnection, user_id: Uuid) -> Result<Option<Self>, UserError> {
         use crate::schema::users::dsl::*;
 
         let user = users.find(user_id).get_result::<User>(conn).optional()?;
 
         Ok(user)
     }
 
-    pub fn create(
-        mut user_data: UserData,
-        conn: &PgConnection,
-    ) -> Result<Self, diesel::result::Error> {
+    pub fn find_by_username(conn: &PgConnection, u: &str) -> Result<Option<User>, UserError> {
+        use crate::schema::users::dsl::*;
+
+        let user = users
+            .filter(username.eq(u))
+            .first::<User>(conn)
+            .optional()?;
+
+        Ok(user)
+    }
+
+    fn username_taken(conn: &PgConnection, u: &str) -> bool {
+        let user = User::find_by_username(conn, u).unwrap();
+        user.is_some()
+    }
```

Comment:
What happens if `User::find_by_username` fails with e.g. a database error? Will this be propagated correctly?

According to [this](https://stackoverflow.com/questions/36362020/what-is-unwrap-in-rust-and-what-is-it-used-for) StackOverflow question it would panic which we probably don't want here.

Reply:
Yes currently it will panic, I can change the return value to a result

Reply:
Done